// @cliqz/adblocker-content/dist/es6/adblocker.js
/*!
 * Copyright (c) 2017-2019 Cliqz GmbH. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
export function getDOMElementsFromMutations(mutations) {
    // Accumulate all nodes which were updated in `nodes`
    const nodes = [];
    for (let i = 0; i < mutations.length; i += 1) {
        const mutation = mutations[i];
        if (mutation.type === 'attributes') {
            nodes.push(mutation.target);
        }
        else if (mutation.type === 'childList') {
            const addedNodes = mutation.addedNodes;
            for (let j = 0; j < addedNodes.length; j += 1) {
                const addedNode = addedNodes[j];
                nodes.push(addedNode);
                if (addedNode.querySelectorAll !== undefined) {
                    const children = addedNode.querySelectorAll('[id],[class],[href]');
                    for (let k = 0; k < children.length; k += 1) {
                        nodes.push(children[k]);
                    }
                }
            }
        }
    }
    return nodes;
}
/**
 * WARNING: this function should be self-contained and not rely on any global
 * symbol. That constraint needs to be fulfilled because this function can
 * potentially be injected in content-script (e.g.: see PuppeteerBlocker for
 * more details).
 */
export function extractFeaturesFromDOM(elements) {
    const ignoredTags = new Set(['br', 'head', 'link', 'meta', 'script', 'style']);
    const classes = new Set();
    const hrefs = new Set();
    const ids = new Set();
    for (let i = 0; i < elements.length; i += 1) {
        const element = elements[i];
        if (element.nodeType !== 1 /* Node.ELEMENT_NODE */) {
            continue;
        }
        if (element.localName !== undefined && ignoredTags.has(element.localName)) {
            continue;
        }
        // Update ids
        const id = element.id;
        if (id) {
            ids.add(id);
        }
        // Update classes
        const classList = element.classList;
        if (classList) {
            for (let j = 0; j < classList.length; j += 1) {
                classes.add(classList[j]);
            }
        }
        // Update href
        const href = element.href;
        if (typeof href === 'string') {
            hrefs.add(href);
        }
    }
    return {
        classes: Array.from(classes),
        hrefs: Array.from(hrefs),
        ids: Array.from(ids),
    };
}
export class DOMMonitor {
    constructor(cb) {
        this.cb = cb;
        this.knownIds = new Set();
        this.knownHrefs = new Set();
        this.knownClasses = new Set();
        this.observer = null;
    }
    queryAll(window) {
        this.handleNewNodes(Array.from(window.document.querySelectorAll('[id],[class],[href]')));
    }
    start(window) {
        if (this.observer === null && window.MutationObserver !== undefined) {
            this.observer = new window.MutationObserver((mutations) => {
                this.handleNewNodes(getDOMElementsFromMutations(mutations));
            });
            this.observer.observe(window.document.documentElement, {
                attributeFilter: ['class', 'id', 'href'],
                attributes: true,
                childList: true,
                subtree: true,
            });
        }
    }
    stop() {
        if (this.observer !== null) {
            this.observer.disconnect();
            this.observer = null;
        }
    }
    handleNewFeatures({ hrefs, ids, classes, }) {
        const newIds = [];
        const newClasses = [];
        const newHrefs = [];
        // Update ids
        for (let i = 0; i < ids.length; i += 1) {
            const id = ids[i];
            if (this.knownIds.has(id) === false) {
                newIds.push(id);
                this.knownIds.add(id);
            }
        }
        for (let i = 0; i < classes.length; i += 1) {
            const cls = classes[i];
            if (this.knownClasses.has(cls) === false) {
                newClasses.push(cls);
                this.knownClasses.add(cls);
            }
        }
        for (let i = 0; i < hrefs.length; i += 1) {
            const href = hrefs[i];
            if (this.knownHrefs.has(href) === false) {
                newHrefs.push(href);
                this.knownHrefs.add(href);
            }
        }
        if (newIds.length !== 0 || newClasses.length !== 0 || newHrefs.length !== 0) {
            this.cb({
                classes: newClasses,
                hrefs: newHrefs,
                ids: newIds,
            });
            return true;
        }
        return false;
    }
    handleNewNodes(nodes) {
        return this.handleNewFeatures(extractFeaturesFromDOM(nodes));
    }
}
/**
 * Wrap a self-executing script into a block of custom logic to remove the
 * script tag once execution is terminated. This can be useful to not leave
 * traces in the DOM after injections.
 */
export function autoRemoveScript(script) {
    // Minified using 'terser'
    return `try{${script}}catch(c){}!function(){var c=document.currentScript,e=c&&c.parentNode;e&&e.removeChild(c)}();`;
    // Original:
    //
    //    try {
    //      ${script}
    //    } catch (ex) { }
    //
    //    (function() {
    //      var currentScript = document.currentScript;
    //      var parent = currentScript && currentScript.parentNode;
    //
    //      if (parent) {
    //        parent.removeChild(currentScript);
    //      }
    //    })();
}
export function injectCSSRule(rule, doc) {
    const parent = doc.head || doc.getElementsByTagName('head')[0] || doc.documentElement;
    if (parent !== null) {
        const css = doc.createElement('style');
        css.type = 'text/css';
        css.id = 'cliqz-adblokcer-css-rules';
        css.appendChild(doc.createTextNode(rule));
        parent.appendChild(css);
    }
}
export function injectScript(s, doc) {
    const script = doc.createElement('script');
    script.type = 'text/javascript';
    script.id = 'cliqz-adblocker-script';
    script.async = false;
    script.appendChild(doc.createTextNode(autoRemoveScript(s)));
    // Insert node
    const parent = doc.head || doc.documentElement;
    if (parent !== null) {
        parent.appendChild(script);
    }
}
